modules/service/src/main/resources/db/migration/V0978__overlap_discount.sql

```diff
@@ -0,0 +1,75 @@
+-- Now that we removed "partner" time and there's no need to distinguish
+-- between "program" vs. "partner", we'll rename the 'c_program_discount'
+-- column to 'c_amount'.
+ALTER TABLE t_time_charge_discount
+  RENAME c_program_discount TO c_amount;
+
+-- A new time charge discount, "overlap" time.  This is time that is not going
+-- to be charged because we switched to some other chargeable visit.  For
+-- example, time spent writing out the last datasets while slewing to another
+-- observation.
+ALTER TYPE e_time_charge_discount_type
+  ADD VALUE 'overlap' AFTER 'nodata';
+
+CREATE TABLE t_time_charge_discount_overlap (
+  c_discount_id    int8             NOT NULL REFERENCES t_time_charge_discount(c_id)    ON DELETE CASCADE,
+  c_observation_id d_observation_id NOT NULL REFERENCES t_observation(c_observation_id) ON DELETE CASCADE,
+
+  PRIMARY KEY (c_discount_id)
+);
+
+-- To keep up with overlaps, we really need a convenient way to get the start
+-- and end time of each visit.  We'll add these to the visit table.
+ALTER TABLE t_visit
+  ADD COLUMN c_start timestamp NULL,
+  ADD COLUMN c_end   timestamp NULL;
+
+-- Initialize the visit times for existing visits.
+UPDATE t_visit v
+   SET c_start = sub.min_ts,
+       c_end   = sub.max_ts
+  FROM (
+    SELECT
+      c_visit_id,
+      min(c_received) AS min_ts,
+      max(c_received) AS max_ts
+    FROM
+      t_execution_event
+    GROUP BY
+      c_visit_id
+  ) sub
+ WHERE v.c_visit_id = sub.c_visit_id;
+
+-- Add an index to make the overlap query faster.
+CREATE INDEX idx_t_visit_site_chargeable_start
+  ON t_visit (c_site, c_chargeable, c_start);
```

Comment:
I think we need an index on the start time at least, but we also require visits to be for the same site and to be chargeable so those are in the overlap query.  Tbh, I'm not sure whether the index should include these as well but it seemed like it?

Reply:
That seems correct if you're joining on all of those columns.